### PR TITLE
Feat: automatic rbf

### PR DESCRIPTION
**Pull request by sander2**
I would like to have automated tests but since this really depends on the bitcoin core specifics maybe the best place would be in the lib..

### CODE REVIEW

File: vault/src/execution.rs [hunk 1]
```diff
@@ -270,9 +271,108 @@ impl Request {
...
+
+            let wait_for_transaction_metadata = btc_rpc.wait_for_transaction_metadata(txid, num_confirmations);
+            futures::pin_mut!(subscription);
+
+            use futures::future::Either;
+            let tx_metadata = match futures::future::select(wait_for_transaction_metadata, subscription.next()).await {
+                Either::Left((result, _)) => result?,
+                Either::Right((Some(Ok((old_fee, new_fee))), _)) => {
+                    tracing::info!("Bumping fee rate from {old_fee} to {new_fee}...");
+                    txid = self.bump_fee(btc_rpc, txid, new_fee).await?;
```
Comment (gregdhill) [inline vault/src/execution.rs:368]: Do we always want to bump the fee as soon as the estimate is higher that the current fee? If it's inexpensive or free I don't see why not but just want to confirm.

**Comment by gregdhill** [comment]
> I would like to have automated tests but since this really depends on the bitcoin core specifics maybe the best place would be in the lib..

As much as I would like to shift the burden of responsibility I do think we're best placed to test this, besides we already have some integration tests [here](https://github.com/interlay/interbtc-clients/blob/master/bitcoin/tests/integration_test.rs).

**Comment by sander2** [comment]
> > I would like to have automated tests but since this really depends on the bitcoin core specifics maybe the best place would be in the lib..
> 
> As much as I would like to shift the burden of responsibility I do think we're best placed to test this, besides we already have some integration tests [here](https://github.com/interlay/interbtc-clients/blob/master/bitcoin/tests/integration_test.rs).

Hmm I didn't remember having those tests. It looks like they aren't run in CI though, so the infrastructure is not setup currently. Do you know we can get a bitcoin regtest node to run in the github test environment?

I do agree that relying on the lib for these tests is not ideal, but we made a decision a long time ago to use the bitcoind simulator instead of using an actual bitcoin node. Do you recall the reason for that? It was a considerable effort to write the simulator so I feel like there must have been a good reason but honestly I can't recall it right now

**Comment by gregdhill** [comment]
> It looks like they aren't run in CI though, and the infrastructure is not setup currently to run them. Do you know how we can get a bitcoin regtest node to run in the github test environment?

Perhaps this was an oversight in the migration from GitLab to GitHub (although it may have actually been Jenkins). @ns212 would be best placed to answer your question though.

> I do agree that relying on the lib for these tests is not ideal, but we made a decision a long time ago to use the bitcoind simulator instead of using an actual bitcoin node. Do you recall the reason for that? It was a considerable effort to write the simulator so I feel like there must have been a good reason but honestly I can't recall it right now

I can't recall the exact reason no, but I think the simulator has its advantages - particularly for unit tests. FWIW I'm not going to block this PR on you adding integration tests but we should definitely test this here and even start to pull other complexities out of the lib.

File: bitcoin/src/lib.rs [hunk 2]
```diff
@@ -964,6 +1019,38 @@ impl BitcoinCoreApi for BitcoinCore {
...
+        let tx = self.rpc.get_raw_transaction(&txid, None)?;
+        let get_tx_result = self.rpc.get_transaction(&txid, None)?;
+
+        // to get from weight to vsize we divide by 4, but round up by first adding 3
+        // Note that we can not rely on tx.get_size() since it doesn't 'discount' witness bytes
+        let vsize = tx
+            .get_weight()
+            .checked_add(3)
+            .ok_or(Error::ArithmeticError)?
+            .checked_div(4)
```
[the remark this answers is not in the document]
Comment (sander2) [reply bitcoin/src/lib.rs:1038]: I just checked - there is `div_ceil`, but it's unstable and feature-gated, and there's no `checked` variant

File: runtime/src/rpc.rs [hunk 3]
```diff
@@ -109,12 +112,17 @@ impl InterBtcParachain {
         let relay_chain_currency_id = currency_constants.get_relay_chain_currency_id()?;
         let wrapped_currency_id = currency_constants.get_wrapped_currency_id()?;
 
+        // low capacity channel since we generally only care about the newest value, so it's ok
+        // if we miss an event
+        let (fee_rate_update_tx, _) = tokio::sync::broadcast::channel(2);
```
[the remark this answers is not in the document]
Comment (sander2) [reply runtime/src/rpc.rs:117]: [Yes](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=e92b890b2641db5313d73e60d5268891)

File: vault/src/execution.rs (hunk 1 above)
Comment (sander2) [reply vault/src/execution.rs:368]: Good point. Bumping fees does not cost the client anything, but there is a minimum increase for the network to accept the new tx. I believe that is set to 1 sat/vbyte, and since our price estimate has a granularity of 1 sat/vbyte, we could always bump. But in case 1 of these assumptions fails to hold (in the future) I will add an additional check for it

Comment (sander2) [reply vault/src/execution.rs]: I decided to just always try and catch the error

File: vault/src/execution.rs [hunk 4]
```diff
@@ -237,16 +239,135 @@ impl Request {
...
+                        .into_inner()
+                        .checked_div(FixedU128::accuracy())
+                        .ok_or(Error::ArithmeticUnderflow)
+                        .and_then(|x| x.try_into().map_err(Into::<Error>::into));
+                    futures::future::ready(ret)
+                })
+                .try_filter_map(|x| async move {
+                    match btc_rpc.fee_rate(txid) {
+                        Ok(current_fee) => {
+                            if x >= current_fee {
```
Comment (gregdhill) [inline vault/src/execution.rs:309]: I am wondering if we should make this configurable, at least if the fee rate increases significantly enough the Vault might prefer to eat the penalty for non-inclusion right?

File: vault/src/execution.rs [hunk 5]
```diff
@@ -237,16 +239,135 @@ impl Request {
...
+                    }
+                    Either::Right((Some(Ok((old_fee, new_fee))), continuation)) => {
+                        tracing::debug!("Attempting to bump fee rate from {old_fee} to {new_fee}...");
+                        match self.bump_fee(btc_rpc, txid, new_fee).await {
+                            Ok(new_txid) => {
+                                tracing::info!("Bumped fee rate. Old txid = {txid}, new txid = {new_txid}");
+                                txid = new_txid;
+                                continue 'outer;
+                            }
+                            Err(Error::BitcoinError(x)) if x.rejected_by_network_rules() => {
```
Comment (gregdhill) [inline vault/src/execution.rs:361]: Will this catch the case that the Vault can not afford the fee increase?

File: vault/src/execution.rs [hunk 6]
```diff
@@ -237,16 +239,135 @@ impl Request {
...
+                    Either::Left((result, _)) => break result?,
+                    Either::Right((None, _)) => return Err(Error::RuntimeError(RuntimeError::ChannelClosed)),
+                    Either::Right((Some(Err(x)), continuation)) => {
+                        tracing::warn!("Received an error from the fee rate subscription: {x}");
+                        // continue with the unchanged fee rate
+                        metadata_fut = continuation;
+                    }
+                    Either::Right((Some(Ok((old_fee, new_fee))), continuation)) => {
+                        tracing::debug!("Attempting to bump fee rate from {old_fee} to {new_fee}...");
+                        match self.bump_fee(btc_rpc, txid, new_fee).await {
```
Comment (gregdhill) [inline vault/src/execution.rs:355]: Is it possible to end up in a sort of "deadlock" if the Vault is always increasing fees, say if we do receive numerous smaller fee estimate bumps on the parachain? (assuming re-submission pushes the new tx to the "back" of the mempool)

**Review by gregdhill — COMMENTED** [review]
The bumping implementation does seem overly complex which makes it hard to follow exactly, but I get the intent. I've left a few comments pertaining to the "always increase" approach which I'm hoping you can provide some more insight on.

File: bitcoin/src/lib.rs [hunk 7]
```diff
@@ -415,11 +436,52 @@ impl BitcoinCore {
...
+
     #[cfg(feature = "regtest-manual-mining")]
-    pub fn mine_block(&self) -> Result<(), Error> {
-        self.rpc
-            .generate_to_address(1, &self.rpc.get_new_address(None, Some(AddressType::Bech32))?)?;
-        Ok(())
+    pub fn mine_block(&self) -> Result<BlockHash, Error> {
+        Ok(self
+            .rpc
+            .generate_to_address(1, &self.rpc.get_new_address(None, Some(AddressType::Bech32))?)?[0])
```
[the remark this answers is not in the document]
Comment (sander2) [reply bitcoin/src/lib.rs:484]: We can always expect one element afaik. But anyway this is only to be used in testing code, so panicking is good here

File: vault/src/execution.rs (hunk 4 above)
Comment (sander2) [reply vault/src/execution.rs:309]: > the Vault might prefer to eat the penalty for non-inclusion right?

I don't understand what you mean here, can you rephrase/elaborate?

File: vault/src/execution.rs (hunk 5 above)
Comment (sander2) [reply vault/src/execution.rs:361]: Hmm that's a good one, probably not. We should add an explicit catch for that I guess

File: vault/src/execution.rs (hunk 6 above)
Comment (sander2) [reply vault/src/execution.rs:355]: I don't see how that'd be possible, no. If `bump_fee` succeeds, the new tx will end up more to the front of the queue because the miners sort by fee. If it fails there'd be no change

File: vault/src/execution.rs (hunk 5 above)
Comment (sander2) [reply vault/src/execution.rs:366]: I decided to not abort at all when getting errors. We'll just wait for the original tx

File: vault/src/execution.rs [hunk 8]
```diff
@@ -239,7 +241,7 @@ impl Request {
     ) -> Result<TransactionMetadata, Error> {
         let fee_rate = self.get_fee_rate(parachain_rpc).await?;
 
-        tracing::debug!("Using fee_rate = {fee_rate}");
+        tracing::debug!("Using fee_rate = {} sat/vByte", fee_rate.0);
```
Comment (gregdhill) [inline vault/src/execution.rs:244]: Nit: could implement `Display` for `SatPerVbyte`

File: vault/src/execution.rs [hunk 9]
```diff
@@ -296,11 +298,11 @@ impl Request {
                 .and_then(|x| {
                     tracing::debug!("Received new inclusion fee estimate {}...", x);
 
-                    let ret: Result<u64, _> = x
+                    let ret: Result<SatPerVbyte, _> = x
                         .into_inner()
                         .checked_div(FixedU128::accuracy())
                         .ok_or(Error::ArithmeticUnderflow)
-                        .and_then(|x| x.try_into().map_err(Into::<Error>::into));
+                        .and_then(|x| x.try_into().map(|x| SatPerVbyte(x)).map_err(Into::<Error>::into));
```
Comment (gregdhill) [inline vault/src/execution.rs:305]:
```suggestion
                        .and_then(|x| x.try_into().map(SatPerVbyte).map_err(Into::<Error>::into));
```

File: vault/src/execution.rs (hunk 4 above)
Comment (gregdhill) [reply vault/src/execution.rs:311]: Okay so this might be an absurd example, but my point is that we may not want to always blindly increase the fee.

So in reference to the fee budget taken from the oracle when the redeem request is made, let's assume we have the following:
- sat/byte is 100
- estimated redeem size is 384 bytes
- fee budget is 38400 sat or 0.000384 BTC = ~$10
- redeem value is 0.1 BTC (total is 0.100384 BTC)

There is no technical limit to the fee rate so let us assume this increases to 3000 sat/byte which the Vault has to cover out-of-pocket:
- vault now pays 1152000 sat (0.01152 BTC = ~$275) - maybe less depending on actual size
- total sent is now 0.11152 BTC

Now assume the Vault decided not to bump the transaction fees:
- slashed 100% value in collateral (e.g. DOT)
- punishment fee (10%) of 0.1 BTC also taken = 0.01 BTC (this may also be increased in the future)
- total lost is 0.11 BTC - lower net loss than bumping the transaction fee

Comment (gregdhill) [reply vault/src/execution.rs:311]: As @sander2 has just reminded me, once the transaction has been sent the Vault anyway risks losing funds if it doesn't bump to include before expiry.

**Review by gregdhill — APPROVED** [review]
LGTM!
